### DISCUSSION TITLE
Rework of explosive damage

### DISCUSSION THREAD

**Discussion by chaosvolt** [cataclysmbn/Cataclysm-BN#2060] [General]
# Is your feature request related to a problem? Please describe.

Currently, explosives have been reworked to be basically useless against terrain. C4 no longer works against mundane walls, even mininukes do very little to terrain. On top of that, the damage also has a severe effect on their usefulness against monsters. Mininukes again are a good example, dealing a level of damage that isn't even enough to kill a hulk.

While obviously fixing their damage would increase how dangerous most explosives are to the player, I feel like this isn't undesirable. Explosives that give shrapnel damage instead of blast damage are ALREADY extremely dangerous to players, NPCs, and monsters alike, so it would be reasonable and consistent for powerful blast-damage explosives to be something you don't want to play around with. Plus, the upper end of blast-focused explosives still deal enough damage to be nasty for the player to face, disproportionately so compared to their usefulness against monsters. That same mininuke that many top-tier monsters will survive would still be powerful enough to kill just about any PC, for example.

# Describe the solution you'd like

One first needs an idea of what type of terrain any given explosive should actually have at least 50% odds of destroying when 1 tile away (i.e. set right next to a targeted wall). I'll list as many suggested values as I can here, any other suggestions and additions will be added to the table as they come up:

Item | Target
--- | ---
Pipebomb | Standard windows
Grenade | Closed doors, possibly boarded up doors
Dynamite | Mundane domestic walls
C4 | Brick walls, possibly concrete
Dynamite bomb | Solid rock
Mininuke | Metal walls, possibly whatever is the absolute strongest wall in the game

I'd suggest we start by referencing DDA's realismic values for explosive power in grams of TNT, and convert it back to damage by dividing it by the damage factor that seems to be in the hardcode: `2/15` (or `0.1333`). This gives us a rough starting point to test on whatever we feel that explosive's desired target should be.

If it can't destroy its desired target terrain reasonably often then buff until it does, if it's tested and found too effective against significantly stronger terrain meant to be a different explosive's primary target, then lower it until the value works best against the target terrain. If possible damage should also favor an amount that makes it reasonably worth the effort of obtaining/crafting it against monsters.

# Describe alternatives you've considered

# Additional context
Add any other context (such as mock-ups, proof of concepts or screenshots) about the feature request here.

**Comment by Relgar** [comment]
~~DeltaEpsilon / @acssimpson has mentioned that the reason why it appears weak is that the explosion is being stopped from being propagated the moment it runs into a tile. After this has been done all that needs to happen should be tweaking damage of the explosives to be outside the max bash damage range, I'm guessing. See below:~~

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/c15e1439b79df596b50e0b71c2e75673c3fcb315/src/explosion.cpp#L169

I would also like to take this opportunity to suggest splitting explosives up by material so that they have differing strengths. Right now you can use multiple explosive types to make an improvised  explosive, but there is no universe where a heap of match stick powder can produce the same power as RDX.

Edit: Turns out this was working as expected, only problem was that this one liner that was quartering the amount of bashing force for some unknown reason aside from the tile the explosive was on. I'm looking at the terrain vs explosive data if this is reverted to full force as we can control for the bashing power with radial damage dropoff.

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/c15e1439b79df596b50e0b71c2e75673c3fcb315/src/explosion.cpp#L181

**Comment by Relgar** [comment]
BFN has also previously implemented radial damage so that dropoff occurs once you're 50% away from the epicenter. This can be rolled into your proposed framework so that walls near the epicenter get blown apart, but the further away they are the less likely this is to happen.

**Comment by Relgar** [comment]
To check the math, here is a selection of the more common walls and doors. The way bashing damage works is that if your bash / explosion damage is between the min and max (inclusive) that'll be your chance / max to bash the terrain. As such, midpoint (rounded down) should represent a 50% chance of destroying the terrain.

Terrain | ID | STR_MIN | MID (RD) | STR_MAX
-- | -- | -- | -- | --
Window | t_window | 3 | 4 | 6
Glass Door | t_door_glass_green_c | 6 | 13 | 20
Wood Door | t_door_c | 8 | 44 | 80
Boarded Up Doors | t_door_boarded | 15 | 47 | 80
Reinforced Wood Door | t_door_glass_green_c | 18 | 59 | 100
Brick Walls | t_brick_wall | 60 | 110 | 160
Normal Wall | t_wall | 30 | 120 | 210
Reinforced Glass Door | t_reinforced_door_glass_c | 40 | 125 | 210
Metal Door | t_door_metal_c | 80 | 165 | 250
Concrete Wall | t_concrete_wall | 70 | 235 | 400
Solid Rock | t_rock | 100 | 250 | 400
Metal Wall | t_wall_metal | 200 | 400 | 600
Resin Wall | t_wall_resin | 250 | 475 | 700

In comparison, here is the list of explosives that we have:


Name | JSON Dmg. | Curr. Dmg | Prop. Dmg | 50% Dmg | Ex. Range | 50% Range | Shrap Dmg. | Shrap. Rng. | Hardest Terrain at Epi | Hardest Terrain at 50%+ Distance
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Can Bomb | 0 | 0 | 0 | 0 | 0 | 0 | 50 | 5 |   |  
EMP Bomb | 50 | 12 | 0 | 0 | 7 | 3 | 0 | 0 |   |  
Improvised   Pipe Bomb | 20 | 5 | 20 | 10 | 4 | 2 | 70 | 8 | Glass Door | Window
Small   Improvised Frag Device | 50 | 12 | 50 | 25 | 6 | 3 | 150 | 20 | Glass Door | Glass Door
Improvised   Grenade | 40 | 10 | 80 | 40 | 3 | 1 | 40 | 6 | Boarded Up Doors | Glass Door
Grenade | 40 | 10 | 80 | 40 | 3 | 1 | 80 | 6 | Boarded Up Doors | Glass Door
Dynamite   Bomb | 70 | 17 | 100 | 50 | 6 | 3 | 100 | 10 | Reinforced Wood Door | Glass Door
Dynamite | 100 | 25 | 210 | 105 | 6 | 3 | 0 | 0 | Reinforced Glass Door | Reinforced Wood Door
Improvised   Demolition Charge | 150 | 37 | 250 | 125 | 6 | 3 | 0 | 0 | Metal Door | Reinforced Wood Door
C-4   Explosive | 200 | 50 | 400 | 200 | 8 | 4 | 0 | 0 | Solid Rock | Brick Walls
ANFO   Charge | 250 | 62 | 500 | 250 | 12 | 6 | 100 | 20 | Solid Rock | Metal Door
Improvised   Barrel Bomb | 200 | 50 | 600 | 300 | 15 | 7 | 80 | 20 | Metal Wall | Metal Door
Mininuke | 300 | 75 | 1000 | 500 | 20 | 10 | 0 | 0 | Resin Wall | Solid Rock



So we can immediately see that currently, a mininuke can't even begin to damage solid rock and onwards. To address this I would first suggest removing the ```bashing_force = force / 4``` line as it is the main culprit in reducing dmg vs terrain. It will be more consistent to use the same damage values in JSON vs terrain as well as vs monsters.

Secondly, to bring this up to the 'expected' things the explosives can do as initially outlined, I would first buff damage to more than 50%, perhaps even closer to 100% as the last thing you envision an expertly planted C-4 bomb to do is failing to blow up the wall it was planted on. The way we would control overdoing the damage would be with the damage dropoff that was initially coded as 50% dmg once the explosion has reached 50% of the distance from the radius. A more nuanced approach could be taken such as 100%, 80%, 60%, 40% corresponding with distance. This would mean, even at the furthest edge, an increase of 160% vs what is currently occurring (400% * 40%).

I'll sleep on this for a bit and tweak the numbers, but please have a look at the initial proposed damage values, keeping in mind that we'll use damage falloff to prevent overbashing.

**Comment by chaosvolt** [comment]
Looks like a good start, yes. Sorry that I haven't had time to try to tinker with this myself, been busy with offline things that keep coming up.

**Comment by Coolthulhu** [comment]
I'm working on a general "terrain being bashed" update, so I'll handle it there.

**Comment by Relgar** [comment]
Some additional bits: there are also ammo effects that need to be considered, and what they are envisioned to do. I just loaded some of these like the RPG-7 and LAW, and they barely dent a normal car. See below values, these will need to be 'upgraded' to actually take out vehicles for some weapons. Maybe an EXPLOSIVE_AP flag:


Name | JSON Dmg. | Curr. Dmg | Prop. Dmg | 50% Dmg | Ex. Range | 50% Range | Shrap Dmg. | Shrap. Rng. | Hardest Terrain at Epi | Hardest Terrain at 50%+ Distance
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
FRAG_SMALL | 0 | 0 | 0 | 0 | 0 | 0 | 30 | 3 |   |  
FRAG | 0 | 0 | 0 | 0 | 0 | 0 | 50 | 5 |   |  
FRAG_BIG | 0 | 0 | 0 | 0 | 0 | 0 | 100 | 7 |   |  
EXPLOSIVE_TINY | 20 | 5 | 40 | 20 | 2 | 1 | 0 | 0 | Glass Door | Glass Door
EXPLOSIVE_SMALL | 40 | 10 | 80 | 40 | 2 | 1 | 0 | 0 | Boarded Up Doors | Glass Door
EXPLOSIVE | 50 | 12 | 250 | 125 | 3 | 1.5 | 0 | 0 | Metal Door | Reinforced Wood Door
EXPLOSIVE_BIG | 70 | 17 | 400 | 200 | 5 | 2.5 | 0 | 0 | Solid Rock | Brick Walls
EXPLOSIVE_HUGE | 100 | 25 | 600 | 300 | 8 | 4 | 0 | 0 | Metal Wall | Metal Door


Ammo   Effect | Ammo | ID
-- | -- | --
EXPLOSIVE_HUGE | 152mm ATGM HEAT | atgm_heat
  | 84x246mm HEDP rocket | 84x246mm_hedp
EXPLOSIVE_BIG | 84x246mm HE rocket | 84x246mm_he
  | PG-7VL 93mm rocket | RPG-7_ammo
  | 66mm HEAT | 66mm_HEAT
EXPLOSIVE | PG-7VR 64mm/105mm rocket | RPG-7_pg7vr
  | TBG-7V 105mm rocket | RPG-7_tbg7v
  | explosive homemade rocket | explosive_hm_rocket
EXPLOSIVE_SMALL | explosive arrow | arrow_exploding
  | explosive crossbow bolt | bolt_explosive
  | Boeing XM-P plasma rifle | plasma_rifle
  | PPA-5 | plasma_gun
EXPLOSIVE_TINY | 20x66mm buckshot | 20x66_shot

Then there is also the problem of holes being randomly generated (as discussed with @Coolthulhu) when the roll hits min-max instead of radiating from the epicenter in a clear, connected, logical way. This affects the pothole-effect you can get from explosions as below (mininuke in a evacuation center basement w/ 700 dmg @ full bash force):

![image](https://user-images.githubusercontent.com/84274761/151367998-fa03e1a6-4cc7-41d3-a686-354979c2341b.png)

**Comment by Coolthulhu** [comment]
> they barely dent a normal car

The problem is with car values, not explosives. They have massive armor.

**Comment by Relgar** [comment]
That's fair, another rebalance to do!

**Comment by Relgar** [comment]
I've updated the numbers - BFN prefers a 50% dmg dropoff so I'm trying to make it so our little nuke can take out anything and still cause a respectable amount of damage at falloff. Please have a look at let me know if the numbers seem off.

**Comment by chaosvolt** [comment]
Important side note though: remember that the damage probably needs a look too, removing the arbitrary quartering of effect on walls is only part of the problem. The way that a grenade can potentially be more lethal to monsters than most high explosives because fragment explosives still have massive damage while blast-focused explosives don't is still a discrepancy.

**Comment by Relgar** [comment]
Yeah I've taken a quick skim of the monsters - aside from the 10k debug monsters, the highest I found was mechs (550) and triffid hearts / trapped tendrils (500). The new damage values for the following can kill them in one shot, if they are at the epicenter:

- Huge Explosions (epicenter, tank shells)
- Mininukes (any range)
- ANFO charge (epicenter)
- Improvised barrel bomb (epicenter)

These explosives are rare enough / difficult enough to make that you wouldn't use them for no reason.

EDIT: You're also right about shrapnel /  frag damage though - those will have to be taken in account together with the blast damage. And I got to rebalance the meme that is 'small improvised frag device'.

**Comment by chaosvolt** [comment]
Seems like that's a good start then, nice.
